tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,294 @@
+      reinterpret_cast<std::int64_t*>(shape_arr->data());
+  tf_dlm_tensor_ctx->tensor.dl_tensor.strides =
+      nullptr;  // NULL indicates tensor is compact and row-majored.
+  tf_dlm_tensor_ctx->tensor.dl_tensor.byte_offset =
+      0;  // TF doesn't handle the strides and byte_offsets here
+  return &tf_dlm_tensor_ctx->tensor;
+}
+
+absl::optional<std::string> DeviceNameFromDlContext(const DLContext& ctx,
+                                                    TF_Status* status) {
+  switch (ctx.device_type) {
+    case DLDeviceType::kDLCPU:
+      return "CPU:0";
+    case DLDeviceType::kDLGPU:
+      return absl::StrCat("GPU:", ctx.device_id);
+    default:
+      return absl::nullopt;
+  };
+}
+TF_DataType TfDataTypeFormDlDataType(const DLDataType& dtype,
+                                     TF_Status* status) {
+  TF_DataType tf_dtype;
+  switch (dtype.code) {
+    case DLDataTypeCode::kDLUInt:
+      switch (dtype.bits) {
+        case 8:
+          tf_dtype = TF_DataType::TF_UINT8;
+          break;
+        case 16:
+          tf_dtype = TF_DataType::TF_UINT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_UINT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_UINT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported UInt bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLInt:
+      switch (dtype.bits) {
+        case 8:
+          tf_dtype = TF_DataType::TF_INT8;
+          break;
+        case 16:
+          tf_dtype = TF_DataType::TF_INT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_INT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_INT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Int bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLFloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_HALF;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_FLOAT;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_DOUBLE;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Float bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLBfloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_BFLOAT16;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported BFloat bits: ", dtype.bits);
+      }
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          "Unsupported Type Codes: ", dtype.code);
+  }
+
+  return tf_dtype;
+}
+
+void DeallocatorWrapperFunc(void* data, size_t len, void* dlmt_vptr) {
```

Comment:
Let's be clear how this is different from `DLManagedTensorDeleter` (one is for DLPack and one is for TF right?).

Reply:
This is an adapter function to match the deallocator function signature in `TFE_NewTensorHandleFromDeviceMemory`
https://github.com/tensorflow/tensorflow/blob/6bf28952981415795ae36784c8c17b08218a4b5c/tensorflow/c/eager/c_api.cc#L1182-L1186